lib/mlxdevm.c

```diff
@@ -183,6 +183,67 @@ mlxdevm_sf_port_add(struct mlxdevm *dl, uint32_t pfnum, uint32_t sfnum)
 	return NULL;
 }
 
+static int cmd_port_dump_cb_to_list(const struct nlmsghdr *nlh, void *data)
+{
+	struct genlmsghdr *genl = mnl_nlmsg_get_payload(nlh);
+	struct nlattr *tb[MLXDEVM_ATTR_MAX + 1] = {};
+	struct port_list_head *head = data;
+	struct mlxdevm_port_list *cur;
+	struct mlxdevm_port *port;
+
+	mnl_attr_parse(nlh, sizeof(*genl), attr_cb, tb);
+	if (!tb[MLXDEVM_ATTR_DEV_BUS_NAME] || !tb[MLXDEVM_ATTR_DEV_NAME] ||
+	    !tb[MLXDEVM_ATTR_PORT_INDEX] || !tb[MLXDEVM_ATTR_PORT_PCI_PF_NUMBER] ||
+	    !tb[MLXDEVM_ATTR_PORT_PCI_SF_NUMBER])
+		return MNL_CB_ERROR;
+
+	cur = calloc(1, sizeof(struct mlxdevm_port_list));
+	TAILQ_INSERT_TAIL(head, cur, entry);
+	port = &cur->port;
+
+	port->port_index = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_INDEX]);
+	port->ndev_ifindex = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_NETDEV_IFINDEX]);
+	port->pfnum = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_PCI_PF_NUMBER]);
+	port->sfnum = mnl_attr_get_u32(tb[MLXDEVM_ATTR_PORT_PCI_SF_NUMBER]);
+	cmd_port_fn_get(tb, port);
+	return MNL_CB_OK;
+}
+
+int mlxdevm_sf_port_dump_list(struct mlxdevm *dl, struct port_list_head *head)
+{
+	struct nlmsghdr *nlh;
+	int err;
+
+	nlh = mnlu_gen_socket_cmd_prepare(&dl->nlg, MLXDEVM_CMD_PORT_GET,
+					  NLM_F_REQUEST | NLM_F_ACK | NLM_F_DUMP);
+
+	mnl_attr_put_strz(nlh, MLXDEVM_ATTR_DEV_BUS_NAME, dl->bus);
```

Comment:
send a follow on patch to do device_handle_set() that wraps these two lines and replace in all the calls.

Reply:
but here we have NLM_F_DUMP, seems not all places need this flag.

Reply:
how about passing flags as a parameter?